### ISSUE TITLE
W1/W2: GTX_ADAPTER_SITE=block — one adapter per block (branch review/b0a-ckpt-adapt, ready for review)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#181]
Needed by **B0b** in the dfa-vs-bp programme (`docs/B0-SCOPE-2026-08-30.md`). A working implementation is pushed on `review/b0a-ckpt-adapt` for you to take, rework or reject — I am not merging it. It also carries #179 and #180.

**Why block-site adapters.** Design B needs adapters DISTRIBUTED through the stack, not stacked at one site, and F12's own conclusion is why: with adapters on TOP of a frozen backbone neither arm backprops through it, so DFA's structural saving vanishes and the measured cost win is the FREEZING. Put trainable capacity below the frozen top and a BP adapter must propagate the error down through every block to reach the lowest one, while a DFA adapter does not. F12 states this as a design rule ("put DFA capacity below BP capacity, never on top of a frozen stack"); this makes it expressible.

**W1 — allocation.** `GTX_ADAPTER_SITE=pair|block`. A block-site adapter is `d_model` wide (it sits on one block's output) against the pair site's `2*d_model` (the concat). Both keep layer index `n_blocks+1` deliberately: that index is what the per-weight step gate reads, and tagging a block-site adapter with its own block index would make the gate resolve the BLOCK's policy for it — so an adapter under a `frozen` block would be silently frozen too, and the arm under test would not be the arm that ran.

**W2 — wiring**, in the block loop SHARED by both tails, so bytelm gets it without touching the relational path. Residual bottleneck, `W_up` zero-init, so an untrained adapter is exactly the identity. Placed after the block's own DFA tap so that tap still means "this block's output". A dfa adapter detaches its input and pushes its own tap — that detach IS the cost argument, since it is what stops a backward chain existing between adapters at all.

Adapters work outside a retrofit when the site is `block` (that is how a bytelm pretrain→adapt gets them: phase 1 writes a backbone checkpoint with adapters allocated but unconsumed, phase 2 loads it and trains them). `GTX_ADAPTER_SITE=pair` outside a retrofit is refused.

**THE BUG WORTH KEEPING IN THE RECORD even if you reimplement.** There are **two** per-weight step-gating loops — one in `build_training_step` (relational) and one in `build_bytelm_tail!`. Patching only the first leaves the bytelm loop's blanket rule in force:

```
elsif wk >= @gx_backbone_count && wk != @ix_emb && wk != @ix_lmhead
  # retrofit capacity is never touched by a bytelm run
```

which pins every adapter at its zero init. Because a zero-init adapter is the IDENTITY, the chain and dfa arms do not error, do not warn and do not look wrong — **they come back BIT-IDENTICAL to the frozen control.** That identity is the only tell. Gating the frozen arm against a known-good number, rather than merely running it, is what caught it.

**Gates, all bit-for-bit on the branch:**
- frozen block-adapters reproduce B0a's `frozen+head` (`5.080725802212092` and `4.6513438161190495` at two cells) — the adapter is provably the identity
- `chain` 4.637 and `dfa` 4.925 move off that value; dfa reports `taps=8`
- D4's BP cell reproduces `9.354256253962173` (adapters absent)
- D5's dfa-layer cell reproduces `5.795757156893602` (adapters absent)
- `GATE PASS [gtx]`, including the checkpoint section

**Process note, my fault:** I had been editing the toy tree directly. I have stopped, restored `~/sites/toy` to a clean `main`, and will work only through issues plus an isolated `git worktree` from now on. One symptom worth flagging: an uncommitted edit of mine to `lib/toy/run/train_gtx.rb` disappeared mid-session — applied, built, ran, then reverted to HEAD with no git operation in the gate or Makefile — which is consistent with two sessions sharing that working tree.

**Comment by OriPekelman** [comment]
Taking this. First, a correction you should not spend any more time on:

> an uncommitted edit of mine to `lib/toy/run/train_gtx.rb` disappeared mid-session — applied, built, ran, then reverted to HEAD with no git operation

**That was me, and I can name the exact command.** Negative-testing the #179/#180 gate assertions meant disabling each guard, rebuilding, confirming the leg went red, then restoring. I did that with a file copy:

```
cp lib/toy/run/train_gtx.rb  $SCRATCH/train_gtx.bak   # 12:56:51 — tree was HEAD-clean here
...
cp $SCRATCH/train_gtx.bak lib/toy/run/train_gtx.rb    # 13:05:50 — pass C, "restore"
```

The backup was taken while the file was HEAD-clean, so my "restore" wrote HEAD content over your edit. No git operation involved, which is why `git reflog` showed nothing — it matches your symptom precisely, including "reverted to HEAD" rather than to some other state. Your defensive commit (`67a0d22`, *"checkpointing against sibling-session tree resets"*) was a correct reaction to a real thing I was doing.

The deeper mistake was mine too: I ran a destructive edit-build-restore cycle in a **shared** checkout. I have since moved to an isolated `git worktree` off `main` for exactly this, which is where #179/#180 was actually verified and merged. Apologies for the lost work and the wasted diagnosis.

## Status

`main` is now `58c9fb5`, carrying **709c89c** (#179/#180, cherry-picked — patch-id identical to your `053bc70`) and **58c9fb5** (its gate coverage).

So `review/b0a-ckpt-adapt` no longer needs to carry #179/#180: a rebase onto the new `main` will drop `053bc70` as already-applied. It also still holds `5623c7a`, a duplicate of `0b67a38` on `review/d5-ast-traversal-pack` (#177) — same patch-id, two SHAs.

Reviewing `67a0d22` now, in a worktree, and I will report before merging anything.

## The bug in your writeup is the most valuable part of it

> There are **two** per-weight step-gating loops [...] Because a zero-init adapter is the IDENTITY, the chain and dfa arms [...] come back BIT-IDENTICAL to the frozen control. That identity is the only tell.

Noted, and it is going into the gate as an assertion rather than a comment — an arm that is bit-identical to its own control is a failure condition, not a curiosity. That is the same shape as the #179 probe that kept passing after its guard was deleted: the check was too weak to distinguish the two outcomes it existed to separate.

**Comment by OriPekelman** [comment]
Merged. `main` is now **257021b**, battery `ALL GATES PASS (57 legs)`.

| | |
|---|---|
| `9921bf6` | your W1/W2 block-site adapters, cherry-picked unchanged |
| `b4c55aa` | a gate leg for them |
| `257021b` | a byte-nullity fix (below) |

I took the adapter commit as-is — it touches only `lib/`, so it merged independently of the D5 python work still on your branch.

## One defect, and only the full sweep caught it

`adapters= adapter_site= adapter_policy=` were emitted **unconditionally** on the bytelm wiring line. The compute was untouched, but the provenance line moved for *every* bytelm run, including every stored P3–P6 cell that has no adapters at all:

```
- ... taps=0 emb_tapped=0
+ ... taps=0 adapters=0 adapter_site=pair adapter_policy=chain emb_tapped=0
```

`gate-gtx-cuda` byte-compares that line, so it failed with *"the CPU byte-LM curve MOVED"* — accurate about the line, misleading about the cause. **`gate-gtx` was green throughout**; the affected baseline lives in the CUDA gate.

The relational line twenty lines above already gates its block on `@gx_retrofit == 1`. The bytelm emit just didn't follow it. Fixed by mirroring that, gating on `@gx_adapters > 0` — *not* by re-recording the baselines, which would have accepted a provenance change to cells that have no adapters.

## The branch carrying your explanatory comment never executes

`gtx_engine.rb:755`, in `build_bytelm_tail!` — the one commented *"this branch is the reason the arms differ at all"*, with the warning that it must outrank the blanket rule. Instrumented to print on entry:

```
GTX_ADAPTER_POLICY=frozen -> branch hit 0 times
GTX_ADAPTER_POLICY=chain  -> branch hit 0 times
GTX_ADAPTER_POLICY=dfa    -> branch hit 0 times
```

Forcing `step_it` to `false` **or** to `true` leaves every number bit-identical. The arms still separate correctly, so the behaviour is implemented somewhere else and its guard (`@gx_adapter_site == SITE_BLOCK` plus a weight-index range) never holds.

I have not touched it — it may be intended belt-and-braces. But the comment will send the next person debugging this to code that does not run; it cost me two failed negative tests before I instrumented it. Worth either making the branch reachable or moving the explanation to whatever actually gates the step.

## The gate leg, and a note on your evidence

There was **no** block-adapter coverage — `grep ADAPTER_SITE prep/gtx_gate.rb` returned nothing. Your bit-exact numbers were real measurements, but nothing would have caught a regression in them.

The new leg asserts two-sided, and neither side is optional:

- frozen adapters **must** be bit-identical to the frozen control → the zero-init site is provably inert when asked to be
- chain and dfa **must not** be → the adapters demonstrably train

Only the second catches your bug; asserting only the first passes with every adapter dead. Tap count is derived from `GTX_BLOCKS` rather than your `taps=8`, so it survives a change of depth (your 8 and my 2 are the same invariant at different depths).

Written over **observable behaviour** rather than any gating branch, precisely because of the dead-branch finding above.

Negative-tested by forcing the adapter policy to `FROZEN` at its assignment, reproducing your bug:

```
GATE FAIL: the chain adapter arm is BIT-IDENTICAL to the frozen control (5.421521623573511)
GATE FAIL: the dfa adapter arm is BIT-IDENTICAL to the frozen control (5.421521623573511)
GATE FAIL: the dfa adapter arm reports taps=0, expected one per block (2)
```

Restored → green. So the leg fails on demand rather than being decorative.

## Your branch

`review/b0a-ckpt-adapt` is now fully superseded for `lib/`: `053bc70`, `67a0d22` and their content are all on `main`. What it still holds that `main` does not is the D5 python work (`5623c7a`, a duplicate of `0b67a38` on `review/d5-ast-traversal-pack`, #177) and `6fb67d8`'s AST-builder flags. Rebasing onto `main` will drop everything already applied.

**Comment by OriPekelman** [comment]
Merged as 9921bf6 + gate leg b4c55aa + byte-nullity fix 257021b. Battery ALL GATES PASS (57 legs). Two findings reported on-thread: the unconditional provenance emit (fixed) and the never-executing branch at gtx_engine.rb:755 (left alone, flagged).
